File: julia/src/priorityqueue.jl
import Base.pop!

const NodeId = UInt32
const NodeIndex = UInt32
const Distance = Int32

struct Visit
	first::Distance
	second::NodeIndex
end

struct PriorityQueue
	tree::Array{Visit, 1}
end


@inline function is_prior(q::PriorityQueue, i::UInt32, j::UInt32)::Bool 
	@inbounds local a = q.tree[i]
	@inbounds local b = q.tree[j]
	a.first < b.first || (a.first == b.first && a.second < b.second)
end

@inline function swap!(q::PriorityQueue, i::UInt32, j::UInt32) 
	@inbounds q.tree[i], q.tree[j] = q.tree[j], q.tree[i]
end


function empty(q::PriorityQueue)::Bool 
	return length(q.tree) == 0
end

function push!(q::PriorityQueue, v::Visit)
	index::UInt32 = length(q.tree) + 1
	push!(q.tree, v)
	while index > 1
		parentIndex::UInt32 = index >> 1
		if is_prior(q, index, parentIndex) 
			swap!(q, index, parentIndex)
			index = parentIndex
		else 
			break
		end
	end
end

function pop!(q::PriorityQueue)::Visit 
	result = q.tree[1]
	top = pop!(q.tree)
	size = length(q.tree)
	if size == 0
		return top
	end
	q.tree[1] = top

	index::UInt32 = 1
	while true
		child::UInt32 = (index  << 1)
		if child > size 
			break
		end
		rIndex::UInt32 = child + 1
		if rIndex <= size && is_prior(q, rIndex, child) 
			child = rIndex
		end
		if is_prior(q, index, child) 
			break
		end
		swap!(q, index, child)
		index = child
	end
	result
end

File: go/src/priorityqueue.go
package main

type PriorityQueue struct {
	tree []Visit
}

func NewPriorityQueue() *PriorityQueue {
	return &PriorityQueue{[]Visit{}}
}

func (q *PriorityQueue) isPrior(i, j int) bool {
	a, b := q.tree[i].first, q.tree[j].first
	return a < b || (a == b && q.tree[i].second < q.tree[j].second)
}

func (q *PriorityQueue) Empty() bool {
	return len(q.tree) == 0
}

func (q *PriorityQueue) swap(i, j int) {
	q.tree[i], q.tree[j] = q.tree[j], q.tree[i]
}

func (q *PriorityQueue) Push(v Visit) {
	index := len(q.tree)
	q.tree = append(q.tree, v)
	for index > 0 {
		parentIndex := ((index + 1) >> 1) - 1
		if q.isPrior(index, parentIndex) {
			q.swap(index, parentIndex)
			index = parentIndex
		} else {
			break
		}
	}
}

func (q *PriorityQueue) Pop() Visit {
	result := q.tree[0]
	size := len(q.tree)
	q.tree[0] = q.tree[size-1]
	size--
	q.tree = q.tree[:size]
	for index := 0; ; {
		child := ((index + 1) << 1) - 1
		if child >= size {
			break
		}
		rIndex := child + 1
		if rIndex < size && q.isPrior(rIndex, child) {
			child = rIndex
		}
		if q.isPrior(index, child) {
			break
		}
		q.swap(index, child)
		index = child
	}
	return result
}
